libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,951 @@
+                    flipped,
+                    length,
+                    key,
+                    checksum,
+                })
+            }
+        } else if ID_ERASE.check(word) {
+            let page = ERASE_PAGE.get(word);
+            ParsedWord::Internal(InternalEntry::Erase { page })
+        } else if ID_CLEAR.check(word) {
+            let min_key = CLEAR_MIN_KEY.get(word);
+            ParsedWord::Internal(InternalEntry::Clear { min_key })
+        } else if ID_MARKER.check(word) {
+            let count = MARKER_COUNT.get(word);
+            ParsedWord::Internal(InternalEntry::Marker { count })
+        } else if ID_REMOVE.check(word) {
+            let key = REMOVE_KEY.get(word);
+            ParsedWord::Internal(InternalEntry::Remove { key })
+        } else if word == ERASED_WORD {
+            return Ok(WordState::Erased);
+        } else {
+            return Ok(WordState::Partial);
+        };
+        if let ParsedWord::Internal(internal) = &valid {
+            if WORD_CHECKSUM.get(word)? != 0 {
+                return Ok(WordState::Partial);
+            }
+            let invalid = match internal {
+                InternalEntry::Erase { page } => *page > self.max_page(),
+                InternalEntry::Clear { min_key } => *min_key > self.max_key(),
+                InternalEntry::Marker { count } => *count > MAX_UPDATES,
+                InternalEntry::Remove { key } => *key > self.max_key(),
+            };
+            if invalid {
+                return Err(StoreError::InvalidStorage);
+            }
+        }
+        Ok(WordState::Valid(valid))
+    }
+
+    /// Builds the storage representation of a user entry.
+    pub fn build_user(&self, key: usize, value: &[u8]) -> Vec<u8> {
+        let length = value.len();
+        let word_size = self.word_size();
+        let footer = self.bytes_to_words(length);
+        let mut result = vec![0xff; (1 + footer) * word_size];
+        result[word_size..][..length].copy_from_slice(value);
+        let mut word = ERASED_WORD;
+        ID_HEADER.set(&mut word);
+        if footer > 0 && is_erased(&result[footer * word_size..]) {
+            HEADER_FLIPPED.set(&mut word);
+            *result.last_mut().unwrap() = 0x7f;
+        }
+        HEADER_LENGTH.set(&mut word, length);
+        HEADER_KEY.set(&mut word, key);
+        HEADER_CHECKSUM.set(&mut word, count_zeros(&result[footer * word_size..]));
+        result[..word_size].copy_from_slice(&word.to_ne_bytes());
+        result
+    }
+
+    /// Sets the padding bit in the first word of a user entry.
+    pub fn set_padding(&self, word: &mut WORD) {
+        ID_PADDING.set(word);
+    }
+
+    /// Sets the deleted bit in the first word of a user entry.
+    pub fn set_deleted(&self, word: &mut WORD) {
+        HEADER_DELETED.set(word);
+    }
+
+    /// Returns the minimum number of words to represent a given number of bytes.
+    ///
+    /// Assumes `bytes + self.word_size()` does not overflow.
```

Comment:
Considering this function is public, I would be tempted to have a longer comment here to justify why this assumption holds for us.

Reply:
It depends what you call public. This is a public function of this module, but this module is private to the crate. So all calls are under control.

Note that all crate-public functions in the store library check their input and return an error. Internal functions (like this one) may have preconditions to avoid useless checks and hindering readability by adding dead code.

I still updated the code to follow the style I have in other places (using a `Preconditions` section).

Reply:
Super. Thanks